### PR TITLE
Add targeted option to RobustDPatch and fix object detection label formats

### LINKED ISSUE #1062: Add targeted version of RobustDPatch
We would like to extend `RobustDPatch` to support targeted attacks.

### PR DESCRIPTION
**Pull request by beat-buesser**
Signed-off-by: Beat Buesser <beat.buesser@ie.ibm.com>

# Description

This pull request adds a targeted attack option to `RobustDPatch`.

Fixes #1062 #1119

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: art/estimators/object_detection/utils.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                                               0 <= y1 < y2 <= H in image scale.
+              - labels (Int64Tensor[N]): the labels for each image
+              - scores (Tensor[N]): the scores or each prediction.
+    :param height: Height of images in pixels.
+    :param width: Width if images in pixels.
+
+    :return: Target values of format `List[Dict[Tensor]]`, one for each input image. The fields of the Dict are as
+             follows:
+
+             - boxes (FloatTensor[N, 4]): the boxes in [x1, y1, x2, y2] format, with 0 <= x1 < x2 <= W and
```
Comment (minhitbk) [inline art/estimators/object_detection/utils.py:74]: I am not sure, but wonder if this should be [y1, x1, y2, x2] when we convert from pt to tf?

File: art/estimators/object_detection/tensorflow_faster_rcnn.py [hunk 2]
```diff
@@ -315,21 +322,25 @@ def _load_model(
...
-
-                  - `boxes`: A list of `nb_samples` size of 2-D tf.float32 tensors of shape [num_boxes, 4] containing
-                    coordinates of the groundtruth boxes. Groundtruth boxes are provided in [y_min, x_min, y_max, x_max]
-                    format and also assumed to be normalized as well as clipped relative to the image window with
-                    conditions y_min <= y_max and x_min <= x_max.
-                  - `labels`: A list of `nb_samples` size of 1-D tf.float32 tensors of shape [num_boxes] containing
-                    the class targets with the zero index assumed to map to the first non-background class.
-                  - `scores`: A list of `nb_samples` size of 1-D tf.float32 tensors of shape [num_boxes] containing
-                    weights for groundtruth boxes.
+        :param y: Targets of format `List[Dict[str, np.ndarray]]`, one for each input image. The fields of the Dict are
```
Comment (minhitbk) [inline art/estimators/object_detection/tensorflow_faster_rcnn.py:332]: I think the description of the format of `y` and  `standardise_output` is confusing here. What I understand from the code is that the format of `y` should depend on the value of `standardise_output`, so not fixed. But the description here seems that format of `y` is fixed as tensorflow and then will be converted inside to pytorch if `standardise_output` is True. So that's confusing.

File: art/estimators/object_detection/tensorflow_faster_rcnn.py [hunk 3]
```diff
@@ -381,16 +397,18 @@ def predict(  # pylint: disable=W0221
 
         :param x: Samples of shape (nb_samples, height, width, nb_channels).
         :param batch_size: Batch size.
-        :param standardise_output: True if output should be standardised. Box coordinates will be normalised to [0, 1]
-                                   and label index will be decreased by 1 to adhere to COCO categories.
+        :param standardise_output: True if output should be standardised to PyTorch format. Box coordinates will be
+                                   scaled from [0, 1] to image dimensions, label index will be increased by 1 to adhere
+                                   to COCO categories and the boxes will be changed to [x1, y1, x2, y2] format, with
+                                   0 <= x1 < x2 <= W and 0 <= y1 < y2 <= H.
         :return: A dictionary containing the following fields:
```
Comment (minhitbk) [inline art/estimators/object_detection/tensorflow_faster_rcnn.py:404]: This line should be removed.

**Review by minhitbk — COMMENTED** [review]
Hi @beat-buesser , I only have a few minor comments. Please solve them and then feel free to merge.

File: art/estimators/object_detection/utils.py (hunk 1 above)
Comment (beat-buesser) [reply art/estimators/object_detection/utils.py:74]: Yes, you are right! I'll fix it.

File: art/estimators/object_detection/tensorflow_faster_rcnn.py (hunk 3 above)
Comment (beat-buesser) [reply art/estimators/object_detection/tensorflow_faster_rcnn.py:404]: Yes.

File: art/estimators/object_detection/tensorflow_faster_rcnn.py (hunk 2 above)
Comment (beat-buesser) [reply art/estimators/object_detection/tensorflow_faster_rcnn.py:332]: I agree.

**Comment by beat-buesser** [comment]
Hi @minhitbk Thank you very much! I have addressed all of you comments with new commits.
